craft_parts/plugins/maven_plugin.py

```diff
@@ -74,6 +77,97 @@ def validate_environment(
                 reason=f"invalid maven version {version!r}",
             )
 
+        system_java_major_version = self._get_system_openjdk_java_major_version()
+        project_java_major_version = self._get_project_java_major_version()
+        # if the project Java version cannot be detected, let it try build anyway.
+        if project_java_major_version is None:
+            return
+
+        if system_java_major_version < project_java_major_version:
+            raise errors.PluginEnvironmentValidationError(
+                part_name=self._part_name,
+                reason="system Java version is less than project Java version."
+                f"system: {system_java_major_version}, project: {project_java_major_version}",
+            )
+
+    def _get_project_java_major_version(self) -> int | None:
+        options = cast(MavenPluginProperties, self._options)
+        maven_executable = "./mvnw" if options.maven_use_mvnw else "mvn"
+        effective_pom_path = Path("./effective.pom")
+        try:
+            self._execute(
+                # Write to file since writing to stdout without maven info logs require sudo, i.e.
+                # sudo mvn help:effective-pom -q -DforceStdout -doutput=/dev/stdout
+                f"{maven_executable} help:effective-pom -Doutput={effective_pom_path}"
+            )
+            project_java_major_version = _parse_project_java_version(effective_pom_path)
+            return _extract_java_version(project_java_major_version)
+        except subprocess.CalledProcessError as err:
+            raise errors.PluginEnvironmentValidationError(
+                part_name=self._part_name,
+                reason=f"failed to generate effective pom: {err}",
+            ) from err
+        finally:
+            effective_pom_path.unlink(missing_ok=True)
+
+
+def _parse_project_java_version(effective_pom_path: Path) -> str | None:
+    """Parse the contents of effective pom XML file.
+
+    The order in which Maven determines the Java version is as follows:
+    1. maven-compiler-plugin's configuration release value
+    2. maven-compiler-plugin's property's release value
+    3. java.version property's value (Spring Boot specific)
+    """
+    tree = defusedxml.ElementTree.parse(effective_pom_path)
+    root = tree.getroot()
+    java_version_element: ET.Element | None = root.find(".//{*}java.version")
+    maven_compiler_release_element: ET.Element | None = root.find(
+        ".//{*}maven.compiler.release"
+    )
+    maven_compiler_plugin_release_element: ET.Element | None = None
+    plugins_element = root.find(".//{*}plugins")
```

Comment:
What does the `*` do here please?

Reply:
This matches the plugins element in all(*) namespaces

Reply:
It would depend on the project, for spring-boot-initializr projects, the namespace is: `http://maven.apache.org/POM/4.0.0`.

Here's something that might be relevant as to why the namespace exist.

> In general it's a good style to add namespace information to your xml file. So it is good style to add those information to your pom file which can be helpful in particular if you are in tools like Eclipse etc.

Reply:
ok, makes sense that we would use `*` because we have no way of knowing what the namespace might be